// @extrawest/node-ts-ocpp/dist/src/impl/Protocol.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var jest_websocket_mock_1 = require("jest-websocket-mock");
var globals_1 = require("@jest/globals");
var Protocol_1 = require("./Protocol");
var events_1 = __importDefault(require("events"));
var ws;
var protocol;
globals_1.jest.useFakeTimers();
globals_1.jest.spyOn(global, "setTimeout");
globals_1.jest.spyOn(console, "error").mockImplementation(function () { return null; });
describe("Test Protocol public function onMessage", function () {
    beforeEach(function () {
        ws = new jest_websocket_mock_1.WS("ws://localhost:1234");
        protocol = new Protocol_1.Protocol(new events_1.default(), ws);
    });
    it("get message from onMessage method successfully with CALL_MESSAGE ", function () { return __awaiter(void 0, void 0, void 0, function () {
        var spyOnMessage, message;
        return __generator(this, function (_a) {
            spyOnMessage = globals_1.jest.spyOn(protocol, "onMessage");
            message = JSON.stringify([
                2,
                "866bc7e1-e80a-4475-9cc6-56a05bb2222c",
                "BootNotification",
                {
                    chargePointVendor: "ocpp-rpc",
                    chargePointModel: "ocpp-rpc",
                }
            ]);
            protocol.onMessage(message);
            expect(spyOnMessage).toBeCalled();
            expect(spyOnMessage).toHaveBeenCalledTimes(1);
            expect(spyOnMessage).toHaveBeenCalledWith(message);
            return [2 /*return*/];
        });
    }); });
    it("get message from onMessage method successfully with CALLRESULT_MESSAGE", function () { return __awaiter(void 0, void 0, void 0, function () {
        var spyOnMessage, message;
        return __generator(this, function (_a) {
            spyOnMessage = globals_1.jest.spyOn(protocol, "onMessage");
            message = JSON.stringify([
                3,
                "866bc7e1-e80a-4475-9cc6-56a05bb2222c",
                {
                    smth: "ocpp-rpc",
                    smthToo: "ocpp-rpc",
                }
            ]);
            protocol.onMessage(message);
            expect(spyOnMessage).toBeCalled();
            expect(spyOnMessage).toHaveBeenCalledTimes(1);
            expect(spyOnMessage).toHaveBeenCalledWith(message);
            return [2 /*return*/];
        });
    }); });
    it("get message from onMessage method successfully with CALLERROR_MESSAGE", function () { return __awaiter(void 0, void 0, void 0, function () {
        var spyOnMessage, message;
        return __generator(this, function (_a) {
            spyOnMessage = globals_1.jest.spyOn(protocol, "onMessage");
            message = JSON.stringify([
                4,
                "866bc7e1-e80a-4475-9cc6-56a05bb2222c",
                "ProtocError",
                "Something happened"
            ]);
            protocol.onMessage(message);
            expect(spyOnMessage).toBeCalled();
            expect(spyOnMessage).toHaveBeenCalledTimes(1);
            expect(spyOnMessage).toHaveBeenCalledWith(message);
            return [2 /*return*/];
        });
    }); });
    it("get error from onMessage method", function () { return __awaiter(void 0, void 0, void 0, function () {
        var spyOnMessage, message;
        return __generator(this, function (_a) {
            spyOnMessage = globals_1.jest.spyOn(protocol, "onMessage");
            message = JSON.stringify([
                2,
            ]);
            protocol.onMessage(message);
            expect(spyOnMessage).toBeCalled();
            expect(spyOnMessage).toHaveBeenCalledTimes(1);
            expect(spyOnMessage).toHaveBeenCalledWith(message);
            expect(console.error).toHaveBeenCalled();
            return [2 /*return*/];
        });
    }); });
    afterEach(function () {
        globals_1.jest.clearAllMocks();
        ws.close();
    });
});
describe("Test Protocol public function callRequest", function () {
    beforeEach(function () {
        ws = new jest_websocket_mock_1.WS("ws://localhost:1234");
        protocol = new Protocol_1.Protocol(new events_1.default(), ws);
    });
    it("Send message with callRequest method successfully", function () {
        var spyOnMessage = globals_1.jest.spyOn(protocol, "callRequest");
        var request = "someRequest";
        var payload = {
            smth: "asfe"
        };
        protocol.callRequest(request, payload);
        expect(protocol.callRequest).toBeCalled();
        expect(spyOnMessage).toHaveBeenCalledTimes(1);
        expect(spyOnMessage).toHaveBeenCalledWith(request, payload);
        expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 10000);
    });
    it("Send message with callRequest method with no response from client", function () {
        var spyOnMessage = globals_1.jest.spyOn(protocol, "callRequest");
        var request = "someRequest";
        var payload = {
            smth: "asfe"
        };
        protocol.callRequest(request, payload);
        expect(protocol.callRequest).toBeCalled();
        expect(spyOnMessage).toHaveBeenCalledTimes(1);
        expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 10000);
    });
    it("Send message with callRequest method with reject", function () { return __awaiter(void 0, void 0, void 0, function () {
        var request, payload;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    protocol.callRequest = globals_1.jest.fn(function () {
                        return new Promise(function (resolve, reject) {
                            reject(function () {
                                throw new Error("error");
                            });
                        });
                    });
                    request = "asdasdas";
                    payload = {
                        smth: "asfe"
                    };
                    return [4 /*yield*/, expect(protocol.callRequest(request, payload)).rejects.toThrow("error")];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    }); });
    afterEach(function () {
        globals_1.jest.clearAllMocks();
        ws.close();
    });
});
