// @extrawest/node-ts-ocpp/dist/src/impl/SchemaValidator.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var SchemaValidator_1 = require("./SchemaValidator");
var BootNotification_json_1 = __importDefault(require("../schemas/BootNotification.json"));
var OcppError_1 = require("./OcppError");
describe("Test SchemaValidator's validate method.", function () {
    var customSchema = {
        "$schema": "http://json-schema.org/draft-06/schema#",
        "$id": "urn:OCPP:Cp:2:2020:3:BootNotificationRequest",
        "comment": "OCPP 2.0.1 FINAL",
        "definitions": {
            "BootReasonEnumType": {
                "description": "This contains the reason for sending this message to the CSMS.\r\n",
                "javaType": "BootReasonEnum",
                "type": "string",
                "additionalProperties": false,
                "format": "date",
                "formatMinimum": "2016-02-06",
                "formatExclusiveMaximum": "2016-12-27",
            },
        },
        "type": "object",
        "additionalProperties": false,
        "properties": {
            "reason": {
                "$ref": "#/definitions/BootReasonEnumType"
            }
        },
        "required": [
            "reason"
        ]
    };
    it("should throw format violation", function () {
        var validator = new SchemaValidator_1.SchemaValidator(BootNotification_json_1.default);
        var t = function () {
            validator.validate({
                chargingStation: {
                    model: 90,
                    vendorName: "sad"
                },
                reason: "sdfsdf",
                test: "sdsd"
            });
        };
        expect(t).toThrow(OcppError_1.ERROR_FORMATIONVIOLATION);
    });
    it("should throw type contstrain violation", function () {
        var validator = new SchemaValidator_1.SchemaValidator(BootNotification_json_1.default);
        var t = function () {
            validator.validate({
                chargingStation: {
                    model: 90,
                    vendorName: "sad"
                },
                reason: "sdfsdf"
            });
        };
        expect(t).toThrow(OcppError_1.ERROR_TYPECONSTRAINTVIOLATION);
    });
    it("should throw property constrain violation for long string", function () {
        var validator = new SchemaValidator_1.SchemaValidator(BootNotification_json_1.default);
        var t = function () {
            validator.validate({
                chargingStation: {
                    model: "sdassssssssssssssssssssssssssssssssssssssssssssssssss",
                    vendorName: "sad"
                },
                reason: "sdfsdf"
            });
        };
        expect(t).toThrow(OcppError_1.ERROR_PROPERTYCONSTRAINTVIOLATION);
    });
    it("should throw protocol error for missing required attribute", function () {
        var validator = new SchemaValidator_1.SchemaValidator(BootNotification_json_1.default);
        var t = function () {
            validator.validate({
                chargingStation: {
                    model: "sda",
                    vendorName: "sad"
                }
            });
        };
        expect(t).toThrow(OcppError_1.ERROR_PROTOCOLERROR);
    });
    it("should throw property violation for invalid enum value", function () {
        var validator = new SchemaValidator_1.SchemaValidator(BootNotification_json_1.default);
        var t = function () {
            validator.validate({
                chargingStation: {
                    model: "sda",
                    vendorName: "sad"
                },
                reason: "sdfsdf"
            });
        };
        expect(t).toThrow(OcppError_1.ERROR_PROPERTYCONSTRAINTVIOLATION);
    });
    it("should throw format violation", function () {
        var validator = new SchemaValidator_1.SchemaValidator(customSchema);
        var t = function () {
            validator.validate({ reason: "123" });
        };
        expect(t).toThrow(OcppError_1.ERROR_FORMATIONVIOLATION);
    });
});
